File: apps/api/internal/service/recalc.go
package service

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/tolga/terp/internal/model"
	"github.com/tolga/terp/internal/repository"
)

// dailyCalcServiceForRecalc defines the interface for daily calculation operations.
type dailyCalcServiceForRecalc interface {
	CalculateDay(ctx context.Context, tenantID, employeeID uuid.UUID, date time.Time) (*model.DailyValue, error)
	RecalculateRange(ctx context.Context, tenantID, employeeID uuid.UUID, from, to time.Time) (int, error)
}

// employeeRepositoryForRecalc defines the interface for employee lookup.
type employeeRepositoryForRecalc interface {
	List(ctx context.Context, filter repository.EmployeeFilter) ([]model.Employee, int64, error)
}

// RecalcError represents a single recalculation failure.
type RecalcError struct {
	EmployeeID uuid.UUID
	Date       time.Time
	Error      string
}

// RecalcResult contains the outcome of a recalculation operation.
type RecalcResult struct {
	ProcessedDays int
	FailedDays    int
	Errors        []RecalcError
}

// monthlyCalcForRecalc defines the interface for monthly calculation after daily recalc.
type monthlyCalcForRecalc interface {
	CalculateMonth(ctx context.Context, employeeID uuid.UUID, year, month int) (*model.MonthlyValue, error)
}

// RecalcService triggers recalculation for employees.
type RecalcService struct {
	dailyCalc    dailyCalcServiceForRecalc
	employeeRepo employeeRepositoryForRecalc
	monthlyCalc  monthlyCalcForRecalc
}

// NewRecalcService creates a new RecalcService instance.
func NewRecalcService(
	dailyCalc dailyCalcServiceForRecalc,
	employeeRepo employeeRepositoryForRecalc,
) *RecalcService {
	return &RecalcService{
		dailyCalc:    dailyCalc,
		employeeRepo: employeeRepo,
	}
}

// SetMonthlyCalcService sets the monthly calculation service for automatic
// monthly recalculation after daily values change.
func (s *RecalcService) SetMonthlyCalcService(monthlyCalc monthlyCalcForRecalc) {
	s.monthlyCalc = monthlyCalc
}

// TriggerRecalc recalculates a single day for one employee.
// After daily calculation, also recalculates the affected month so that
// monthly evaluation values (flextime balance, totals) stay in sync.
func (s *RecalcService) TriggerRecalc(ctx context.Context, tenantID, employeeID uuid.UUID, date time.Time) (*RecalcResult, error) {
	_, err := s.dailyCalc.CalculateDay(ctx, tenantID, employeeID, date)
	if err != nil {
		return &RecalcResult{
			ProcessedDays: 0,
			FailedDays:    1,
			Errors: []RecalcError{
				{EmployeeID: employeeID, Date: date, Error: err.Error()},
			},
		}, err
	}

	// Recalculate the affected month so monthly values reflect the daily change
	if s.monthlyCalc != nil {
		_, _ = s.monthlyCalc.CalculateMonth(ctx, employeeID, date.Year(), int(date.Month()))
	}

	return &RecalcResult{ProcessedDays: 1, FailedDays: 0}, nil
}

// TriggerRecalcRange recalculates a date range for one employee.
func (s *RecalcService) TriggerRecalcRange(ctx context.Context, tenantID, employeeID uuid.UUID, from, to time.Time) (*RecalcResult, error) {
	count, err := s.dailyCalc.RecalculateRange(ctx, tenantID, employeeID, from, to)
	if err != nil {
		// Calculate which day failed based on count
		failedDate := from.AddDate(0, 0, count)
		totalDays := int(to.Sub(from).Hours()/24) + 1
		return &RecalcResult{
			ProcessedDays: count,
			FailedDays:    totalDays - count,
			Errors: []RecalcError{
				{EmployeeID: employeeID, Date: failedDate, Error: err.Error()},
			},
		}, err
	}
	return &RecalcResult{ProcessedDays: count, FailedDays: 0}, nil
}

// TriggerRecalcBatch recalculates a date range for multiple employees.
// Continues processing on individual errors.
func (s *RecalcService) TriggerRecalcBatch(ctx context.Context, tenantID uuid.UUID, employeeIDs []uuid.UUID, from, to time.Time) *RecalcResult {
	result := &RecalcResult{}

	for _, empID := range employeeIDs {
		empResult, err := s.TriggerRecalcRange(ctx, tenantID, empID, from, to)
		result.ProcessedDays += empResult.ProcessedDays
		result.FailedDays += empResult.FailedDays
		if err != nil {
			result.Errors = append(result.Errors, empResult.Errors...)
		}
	}

	return result
}

// TriggerRecalcAll recalculates a date range for all active employees in a tenant.
func (s *RecalcService) TriggerRecalcAll(ctx context.Context, tenantID uuid.UUID, from, to time.Time) (*RecalcResult, error) {
	// Get all active employees
	isActive := true
	filter := repository.EmployeeFilter{
		TenantID: tenantID,
		IsActive: &isActive,
	}

	employees, _, err := s.employeeRepo.List(ctx, filter)
	if err != nil {
		return nil, err
	}

	// Extract employee IDs
	employeeIDs := make([]uuid.UUID, len(employees))
	for i, emp := range employees {
		employeeIDs[i] = emp.ID
	}

	return s.TriggerRecalcBatch(ctx, tenantID, employeeIDs, from, to), nil
}

File: src/lib/services/recalc.ts
/**
 * RecalcService
 *
 * Orchestrates recalculation of daily and monthly values for employees.
 * Acts as a coordination layer between DailyCalcService and MonthlyCalcService.
 *
 * Ported from Go: apps/api/internal/service/recalc.go (146 lines)
 *
 * Dependencies:
 * - ZMI-TICKET-234: DailyCalcService (daily time calculations)
 * - ZMI-TICKET-238: MonthlyCalcService (monthly aggregations)
 */

import type { PrismaClient } from "@/generated/prisma/client"
import { mapWithConcurrency } from "@/lib/async"
import { DailyCalcService } from "./daily-calc"
import type { TenantCalcCache } from "./daily-calc.context"
import { loadTenantCalcCache } from "./daily-calc.context"
import { MonthlyCalcService } from "./monthly-calc"
import type { RecalcResult } from "./recalc.types"

export class RecalcService {
  private dailyCalcService: DailyCalcService
  private monthlyCalcService: MonthlyCalcService

  constructor(
    private prisma: PrismaClient,
    dailyCalcService?: DailyCalcService,
    monthlyCalcService?: MonthlyCalcService,
    tenantId?: string,
  ) {
    this.dailyCalcService = dailyCalcService ?? new DailyCalcService(prisma)
    this.monthlyCalcService =
      monthlyCalcService ?? new MonthlyCalcService(prisma, tenantId!)
  }

  /**
   * Recalculates a single day for one employee.
   * After daily calculation, also recalculates the affected month so that
   * monthly evaluation values (flextime balance, totals) stay in sync.
   */
  async triggerRecalc(
    tenantId: string,
    employeeId: string,
    date: Date,
  ): Promise<RecalcResult> {
    try {
      await this.dailyCalcService.calculateDay(tenantId, employeeId, date)
    } catch (err) {
      return {
        processedDays: 0,
        failedDays: 1,
        errors: [
          {
            employeeId,
            date,
            error: err instanceof Error ? err.message : String(err),
          },
        ],
      }
    }

    // Recalculate the affected month so monthly values reflect the daily change
    try {
      await this.monthlyCalcService.calculateMonth(
        employeeId,
        date.getUTCFullYear(),
        date.getUTCMonth() + 1,
      )
    } catch {
      // Monthly recalc is best-effort (matches Go: `_, _ = s.monthlyCalc.CalculateMonth(...)`)
    }

    return { processedDays: 1, failedDays: 0, errors: [] }
  }

  /**
   * Recalculates a date range for one employee.
   * Does NOT trigger monthly recalculation (matches Go behavior).
   */
  async triggerRecalcRange(
    tenantId: string,
    employeeId: string,
    from: Date,
    to: Date,
    tenantCache?: TenantCalcCache,
  ): Promise<RecalcResult> {
    try {
      const { count } = await this.dailyCalcService.calculateDateRange(
        tenantId,
        employeeId,
        from,
        to,
        tenantCache,
      )
      return { processedDays: count, failedDays: 0, errors: [] }
    } catch (err) {
      // Calculate total expected days for failure reporting
      const totalDays =
        Math.floor(
          (to.getTime() - from.getTime()) / (24 * 60 * 60 * 1000),
        ) + 1

      return {
        processedDays: 0,
        failedDays: totalDays,
        errors: [
          {
            employeeId,
            date: from,
            error: err instanceof Error ? err.message : String(err),
          },
        ],
      }
    }
  }

  /**
   * Recalculates a date range for multiple employees.
   * Continues processing on individual errors.
   */
  async triggerRecalcBatch(
    tenantId: string,
    employeeIds: string[],
    from: Date,
    to: Date,
  ): Promise<RecalcResult> {
    // Pre-load tenant-level data once (holidays + settings)
    const tenantCache = await loadTenantCalcCache(this.prisma, tenantId, from, to)

    const empResults = await mapWithConcurrency(
      employeeIds,
      5,
      (empId) => this.triggerRecalcRange(tenantId, empId, from, to, tenantCache),
    )

    const result: RecalcResult = {
      processedDays: 0,
      failedDays: 0,
      errors: [],
    }
    for (const empResult of empResults) {
      result.processedDays += empResult.processedDays
      result.failedDays += empResult.failedDays
      result.errors.push(...empResult.errors)
    }

    return result
  }

  /**
   * Recalculates a date range for all active employees in a tenant.
   */
  async triggerRecalcAll(
    tenantId: string,
    from: Date,
    to: Date,
  ): Promise<RecalcResult> {
    // Get all active employees
    const employees = await this.prisma.employee.findMany({
      where: {
        tenantId,
        isActive: true,
        deletedAt: null,
      },
      select: { id: true },
    })

    // Extract employee IDs
    const employeeIds = employees.map((emp) => emp.id)

    return this.triggerRecalcBatch(tenantId, employeeIds, from, to)
  }
}
